File: cpp/src/Sort.cpp

#include <sort/Sort.hpp>

namespace {

//// Heap sort

template<typename T>
void heapify(std::vector<T>& v, std::size_t n, std::size_t index) {
    auto largest = index;
    const auto left = 2 * index + 1;
    const auto right = 2 * largest + 2;

    if (left < n && v[left] > v[largest]) {
        largest = left;
    }

    if (right < n && v[right] > v[largest]) {
        largest = right;
    }

    if (largest != index) {
        std::swap(v[index], v[largest]);
        heapify(v, n, largest);
    }
}

template<typename T>
void heapSort(std::vector<T>& v) {
    if (v.empty()) {
        return;
    }

    const auto n = v.size();
    for (int64_t i = n / 2 - 1; i >= 0; --i) {
        heapify(v, n, static_cast<std::size_t>(i));
    }

    for (std::size_t i = n - 1; i > 0; --i) {
        std::swap(v[0], v[i]);
        heapify(v, i, 0);
    }
}

//// Quick sort

template<typename T>
std::size_t partition(std::vector<T>& v, std::size_t low, std::size_t high) {
    const auto& pivot = v[high];
    auto i = low;
    for (std::size_t j = low; j < high; ++j) {
        if (v[j] <= pivot) {
            std::swap(v[i], v[j]);
            ++i;
        }
    }
    std::swap(v[i], v[high]);
    return i;
}

template<typename T>
void quickSort(std::vector<T>& v, std::size_t low, std::size_t high) {
    if (low >= high || low > v.size() - 1) {
        return;
    }

    const auto pivot = partition(v, low, high);
    quickSort(v, low, pivot - 1);
    quickSort(v, pivot + 1, high);
}



template<typename T>
void quickSort(std::vector<T>& v) {
    if (v.empty()) {
        return;
    }

    quickSort(v, 0, v.size() - 1);
}

//// Insertion sort

template<typename T>
void insertionSort(std::vector<T>& v) {
    if (v.empty()) {
        return;
    }

    for (std::size_t i = 1; i < v.size(); ++i) {
        const auto key = std::move(v[i]);
        std::size_t j = i;

        while (j > 0 && v[j - 1] > key) {
            v[j] = std::move(v[j - 1]);
            --j;
        }
        v[j] = std::move(key);
    }
}

} // namespace

namespace algorithm {

namespace heap {

void sort(std::vector<int>& container) {
    heapSort(container);
}

void sort(std::vector<float>& container) {
    heapSort(container);
}

void sort(std::vector<std::string>& container) {
    heapSort(container);
}

} // namespace heap

namespace quick {

void sort(std::vector<int>& container) {
    quickSort(container);
}

void sort(std::vector<float>& container) {
    quickSort(container);
}

void sort(std::vector<std::string>& container) {
    quickSort(container);
}

}

namespace insertion {

void sort(std::vector<int>& container) {
    insertionSort(container);
}

void sort(std::vector<float>& container) {
    insertionSort(container);
}

void sort(std::vector<std::string>& container) {
    insertionSort(container);
}

}

} // namespace algorithm

File: py/sort.py
#
# insertion sort
#

def insertion_sort(data):
    if not data:
        return
    
    for i in range(len(data)):
        key = data[i]
        j = i

        while j > 0 and data[j - 1] > key:
            data[j] = data[j - 1]
            j -= 1
        data[j] = key

#
# quick sort
#

def partition(data, low, high):
    pivot = data[high]
    i = low
    for j in range(low, high):
        if data[j] <= pivot:
            data[i], data[j] = data[j], data[i]
            i += 1
    data[i], data[high] = data[high], data[i]
    return i

def quick_sort(data):
    if not data:
        return
    _quick_sort(data, 0, len(data) - 1)


def _quick_sort(data, low, high):
    if low >= high or low > len(data) - 1 or high > len(data) - 1: 
        return
    
    pivot = partition(data, low, high)
    quick_sort(data, low, pivot - 1)
    quick_sort(data, pivot + 1, high)

# 
# heap sort
#

def heapify(data, n , index):
    largest = index
    left = 2 * index + 1
    right = 2 * index + 2

    if left < n and data[left] > data[largest]:
        largest = left

    if right < n and data[right] > data[largest]:
        largest = right

    if largest != index:
        data[index], data[largest] = data[largest], data[index]
        heapify(data, n, largest)                

def heap_sort(data):
    if not data:
        return
    
    n = len(data)
    for i in range(n // 2 - 1, -1, -1):
        heapify(data, n, i)

    for i in range(n - 1, 0, -1):
        data[0], data[i] = data[i], data[0]
        heapify(data, i, 0)
